src/AggregateFunctions/AggregateFunctionEntropy.h

```diff
@@ -23,18 +23,12 @@ struct EntropyData
 {
     using Weight = UInt64;
 
-    using HashingMap = HashMap<
-        Value, Weight,
-        HashCRC32<Value>,
-        HashTableGrower<4>,
-        HashTableAllocatorWithStackMemory<sizeof(std::pair<Value, Weight>) * (1 << 3)>>;
+    using HashingMap = HASH_TABLE_WITH_STACK_MEMORY(HashMap, Value, Weight,
```

Comment:
Ok.

My only concern is that this macro looks ugly. Couldn't we do something better?
Maybe leave type definitions as is (it will have redundand param, but the check will work).

Reply:
> leave type definitions as is

I started with this, but then you have to write out the full cell type, and it's even uglier.
Another bad thing about this macro is that it instantiates the template twice -- first time with the default allocator and second time with stack one. It probably compiles longer than it should and can lead to weird effects.
I experimented with variadic templates but couldn't make it work. Maybe we should ask our C++ experts on the TG chat.

Reply:
Just my opinion (need to ask more developers): writing cell type twice is less bad than using a macro.

Reply:
The full type looks like this:

```
ClearableHashMap<
    StringRef, UInt64, StringRefHash,
    HashTableGrower<initial_size_degree>,
    HashTableAllocatorWithStackMemory<
        (1ULL << initial_size_degree)
        * sizeof(ClearableHashMapCell<StringRef, UInt64, StringRefHash>)
    >
>;
```

It's very verbose but OK, I can manage to type it a couple of times. The thing that really bothers me is not its visual ugliness or verbosity, but the amount of knowledge required to write this.

Besides usual key/value/hash parameters, you also have to know:
* that you have to use some 'grower' that has a 'degree', and the initial number of elements is 1 << degree;
* the initial size of hash table, when you use a standard grower, is 1 << size_degree;
* then you have to multiply it by the cell size, you have to know which cell type the table uses, and get all its parameters right.

No wonder people are struggling with this. We even know the right size and can write an assert for it, so why make a human jump through all these hoops? I just want to write: give me a hash table of type ClearableHashMap that has these key/value/hashes, and also uses 512 bytes of inline memory.

Reply:
We can create template typedef of ClearableHashMap that will take HashTableAllocatorWithStackMemory as template template parameter
and substitute the size of the cell automatically.

Reply:
Or better - just make template typedef `ClearableHashMapWithStackMemory`.